### ISSUE TITLE
postgresql / sqlite / mysql /oracle regular expression operators

### ISSUE THREAD

**Issue by sqlalchemy-bot** [sqlalchemy/sqlalchemy#1390]
**Migrated issue, originally created by Anonymous**

match operator not implemented 

in sqlalchemy/databases/postgres.py 727
just add 
      sql_operators.match_op: lambda x, y, escape=None: '%s ~ %s' % (x, y) + (escape and ' ESCAPE \'%s\'' % escape or ''),

and it works for me 
CAVEAT sql might crash if improper regexp syntax entered.

+ missing imatch / notmatch ....

(these matching functions are still used with same syntax)
http://www.postgresql.org/docs/7.4/interactive/functions-matching.html

----------------------------------------
Attachments: [1390.patch](../wiki/imported_issue_attachments/1390/1390.patch)

----------------------------------------
> Edit 2020-07-10

Coping this recap from https://github.com/sqlalchemy/sqlalchemy/issues/5447#issue-654578670

The call `column.regex_match('[a-z]*')` would then evaluate to:
* [Postgres](https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-POSIX-REGEXP): `column ~ "[a-z]*"`
* [SQLite](https://www.sqlite.org/lang_expr.html#the_like_glob_regexp_and_match_operators): `column REGEXP "[a-z]*"`
* [MySQL](https://dev.mysql.com/doc/refman/8.0/en/regexp.html#operator_regexp), [MariaDB](https://mariadb.com/kb/en/regexp/): `column REGEXP "[a-z]*"`
* [Oracle](https://docs.oracle.com/cd/B12037_01/server.101/b10759/conditions018.htm#SQLRF00501): `REGEXP_LIKE(column, "[a-z]*")`
* MSSQL: throw exception

**Comment by zzzeek** [comment]
note from #5447 this is also possible on other backends:

```
Postgres: column ~ "[a-z]*"
SQLite: column REGEXP "[a-z]*"
MySQL: column REGEXP "[a-z]*"
Oracle: REGEXP_LIKE(column, "[a-z]*")


```

**Comment by TMiguelT** [comment]
I'm happy to look into this. Can you explain how the [Gerrit patch](https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/101) relates to this?

**Comment by zzzeek** [comment]
I would say you can look at that patch to see if there's anything relevant to use.  it proposes new operators for PostgreSQL's version of CHAR, VARCHAR, and TEXT, but if these operators are generally available on many backends, it would instead be made as part of the base String class in sqlalchemy/sql/sqltypes.py .

as far as contributing code we use the normal pull request process.  I merge code through gerrit but outside contributors don't need to worry about it, we synchronize to your pull request.

**Comment by zzzeek** [comment]
also I would note we've written various kinds of background on helping with development at https://www.sqlalchemy.org/develop.html.  

This is certainly an important feature that we want to do, so if you arent able to work on it, no worries it will get done eventually.

**Comment by TMiguelT** [comment]
Great! Thanks for the info, I'll have a look. I'd certainly like for the regex operators to be available generally, but I'm worried that tiny differences in the regex implementations will make this tricky. For instance, I also would like regex flags (e.g. case insensitivity) to be implemented, but some backends, like SQLite, don't implement flags at all. How do you think I should handle this?

**Comment by TMiguelT** [comment]
And some implementations provide regex-related functions like `regex_replace` and `regex_findall` that probably don't make sense to be standardised. I suppose these can go into the dialect columns only, and not the base string class.

**Comment by zzzeek** [comment]
yeah those are all reasons this kind of feature moves slowly because it's easy to get it wrong.

Usually we make a grid of features and figure out what the best "defaults" are, then for optional things we add those options which are not implemented for dialects that don't support it.  

also for "replace" / "findall" we can omit those for now, i think people mostly want to have a WHERE clause matching things.

**Comment by zzzeek** [comment]
i would also note that we don't try to automate regular comparisons as far as case sensitivity, some DBs are case sensitive for comparisons by default some are not, sometimes it is based on the collation settings.   so we probably would deliver the regular expression operator in its default form for each database and only try to provide case sensitive/insensitive with additional options.

**Comment by TMiguelT** [comment]
Okay, here's a feature matrix:

Feature | Postgres | MySQL | SQLite | Oracle | MSSQL
--- | --- | --- | --- | --- | ---
Does `column` match `regex`? | [`column ~ regex`](https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-POSIX-REGEXP) | [`column REGEXP regex`](https://dev.mysql.com/doc/refman/8.0/en/regexp.html#operator_regexp) | [`column REGEXP regex`](https://www.sqlite.org/lang_expr.html#the_like_glob_regexp_and_match_operators) (requires plugin, but has syntax support) |  [`REGEXP_LIKE(column, regex)`](https://docs.oracle.com/cd/B12037_01/server.101/b10759/conditions018.htm) | n/a
Does `column` match `regex` using `flags`? | [`column ~ (?flags)regex`](https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-POSIX-REGEXP) | [`REGEXP_LIKE(column, regex, flags)`](https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-like)  | n/a | [`REGEXP_LIKE(column, regex, flags)`](https://docs.oracle.com/cd/B12037_01/server.101/b10759/conditions018.htm) | n/a
Replace `pattern` with `replacement` in `column` | [`regexp_replace(column, regex, replacement)`](https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-POSIX-REGEXP) | [`REGEXP_REPLACE(column, regex, replacement)](https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-replace) | n/a (can be loaded as a custom function, but no syntax exists for this) | [`REGEXP_REPLACE(column, regex, replacement)`](https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions130.htm) | n/a

Based on this, I think it makes sense to add the first row ("does `column` match `regex`") as a core string-column function, but I'm undecided on the other two features. It would still be helpful to users to have them available in a database-independent manner, event though they are less well-supported.

**Comment by zzzeek** [comment]
cc @jvanasco @gordthompson  @CaselIT

**Comment by jvanasco** [comment]
First question that pops to mind: Does everything implement PCRE or do they use other flavors of regex?

**Comment by CaselIT** [comment]
Also what would the api look like? Do we try to go for a single function that does all tree cases, like `regepx(pattern, flags=None, replacement=None)` or two cases `regepx(pattern, flags=None)` and `regepx_replace(pattern, replacement)`?

I think the second one would make more sense. Also do replacements support flags?

**Comment by TMiguelT** [comment]
> First question that pops to mind: Does everything implement PCRE or do they use other flavors of regex?

This is a bit confusing to work out, but the documentation says:

* Postgres supports POSIX 1003.2 regex, including extended REs and basic REs
* MySQL now uses Unicode Regular Expressions, which I think are a superset of POSIX
* SQLite depends on the plugin, but the popular [`sqlite3-pcre)`](https://github.com/ralight/sqlite3-pcre) uses PCRE
* Oracle uses "POSIX standard Extended Regular Expression (ERE)"

A comparison table is available [here](https://gist.github.com/CMCDragonkai/6c933f4a7d713ef712145c5eb94a1816#feature-comparison). In summary it seems that not all the backends have the same dialect, but they're fairly similar at least. Should we output the regex provided by the user verbatim, and warn them to use a portable subset of regex, or should we attempt to convert the regex to a backend-compatible version at execution time? (this sounds hard).

**Comment by CaselIT** [comment]
> Should we output the regex provided by the user verbatim, and warn them to use a portable subset of regex, or should we attempt to convert the regex to a backend-compatible version at execution time? (this sounds hard).

I think that a note on the documentation is enough in this case. It would be a lot of work to try and convert the regexp so I think that we should just render the input verbatim and let the db error if it is not correct

**Comment by TMiguelT** [comment]
> Also what would the api look like? Do we try to go for a single function that does all tree cases, like regepx(pattern, flags=None, replacement=None) or two cases regepx(pattern, flags=None) and regepx_replace(pattern, replacement)?

Certainly replace and match should be separate functions. But what I'm not sure about is the flags. We could have `flags` as a separate argument for both of these methods, but then what happens when we execute `column.regexp(pattern, flags)` on SQLite? Do we throw an error? Do we simply remove the flags and run it anyway? If we do this, the match behaviour is going to be very different on SQLite.

> Also do replacements support flags?

I have updated the table above. All the backends that support replacement seem to support flags.

**Comment by jvanasco** [comment]
> I think that a note on the documentation is enough in this case.

Absolutely. I brought this up in case any of the backing implementations were way off from one another.  If that were the case, it might be worth not supporting regex  natively on that platform. That doesn’t not seem applicable to these options though.

**Comment by zzzeek** [comment]
do the flags for each backend mean the same thing?    if there are flags that mean something only on one backend then we do something like mysql_someflag .    we try not to do that if we don't have to.

**Comment by jvanasco** [comment]
yes and now. i'll make a matrix.

**Comment by TMiguelT** [comment]
I think they're sufficiently similar that we should try to unify them. For example, `i` in all backends means case-insensitive, `c` means case-sensitive and `m` means multiline. But you need `nm` in postgres to have equivalent multine behaviour to `m` in MySQL and Oracle.

**Comment by TMiguelT** [comment]
Since its really only those two/three settings that are commonly controlled by flags, we could just add some boolean arguments to the match function that abstract away the flags themselves, for example `def regexp(pattern, case_insensitive=False, dot_newline=False, multiline=False)`

**Comment by zzzeek** [comment]
I don't think we've ever had a new feature with such a great roundtable discussion going on to make sure we get it right (and that I can other things done while it happens! :)  )  thanks all !!!

**Comment by CaselIT** [comment]
I'm 👎 on the boolean argument to specify the flags, since they are hard to change/scale. In case of regexp is not unusual to specify the flag as a string, so I would propose to just use a string that is rendered in the sql.

**Comment by TMiguelT** [comment]
In theory I would prefer that too, but as discussed above, the multiline behaviour is quite different between DBMS. If you called `column.regexp('pattern', flags='m')`, it would invoke "`.`-matches-newline" **and** "`$/^` matches on linebreaks" for Postgres, but only `$/^` matches on linebreaks" on all other backends.

I can't think of a better way to provide abstraction without using boolean arguments. Python's `re` module uses bitwise flags for this behaviour, but that's less user-friendly, if a little more scalable.

**Comment by CaselIT** [comment]
I believe that the same argument could be made for `'pattern'`. I don't think we can assume that they are the same in all backends, so I'm not sure if it would be a pro to have the flag that are backend independent but the pattern is not.

Of the two I would prefer something like `column.regexp('pattern', flags='m', postgres_flags='nm', somedb_pattern='patern')` to have the ability of specifying backend dependent flags.

Also regarding flags, if ` multiline=True`  gets translated to 'mn' in pg, expect someone to want only 'm' or only 'n'.

That said, I'm not sure we need to make it backend independent, at leas as a first implementation.

**Comment by jvanasco** [comment]
what if the regex operator just accepts the core arguments, and there is an ancillary operator/factory for customizing the operations via a dict or callable... so the experience is more inline with the custom-compiles? 

for example...

   column.regexp('pattern', flags='m')

and 

   column.regexp_factory({"postgresql": ['pattern', flags='m'],
						  "mysql": ['pattern', flags='m'],
						   })

or

   def regexp_factory(backend):
        return pattern, flags
   column.regex(callable=regexp_factory)

**Comment by TMiguelT** [comment]
> I believe that the same argument could be made for 'pattern'

Not exactly, there is a common subset of all these regex implementations, which I think would cover 95% of cases. The same isn't true for flags, since the multiline flags are quite different

Both of the `postgres_flags` suggestions make sense, but I worry that we aren't providing enough abstraction if we do that, which I would have thought was one of the main purposes of SQLAlchemy. The average user says "I want to match this column in multiline mode", and doesn't want to have to look up the database manuals for every single DBMS in order to achieve this portably.

**Comment by zzzeek** [comment]
since we have a lot of flags and things here, the interface might best be done in the style of "variant",a little tricky to get it right, but it looks like:

     col.regexp(pattern, flags).with_variant("mysql", pattern, flags) ...

i haven't looked closely at the discussion here so don't fret if I'm missing the point entriely

**Comment by CaselIT** [comment]
Is there an helper/mixin for the `with_variant` functionality? I believe it is used already in some apis

**Comment by zzzeek** [comment]
it's very specific to TypeEngine.      the approach can potentially be emulated.

for the moment we could move forward with this without doing all that, however.  most people using regex are probably targeting just one database.

**Comment by TMiguelT** [comment]
As I've mentioned elsewhere, I think it's quite important that we *do* come up with a portable solution. The whole reason I've investigated this issue is actually because I wanted to do a regex match using the same codebase on Postgres and SQLite. If I wanted to only support Postgres I'd use the `op("~")`.

I don't mind the `with_variant()`. The only problem, as I've said, is that it still involves actually knowing the differences between pattern/flag behaviour on each DBMS which your average user isn't aware of. Still, doing this and linking to the appropriate documentation would be a good start.

**Comment by jvanasco** [comment]
@zzzeek's`with_variant` style does the same as my idea, but it more SQLAlchemy-like, so I'm +1 for it.

@TMiguelT writing a regex that works across multiple database backends is really outside the scope of what SQLAlchmey does or promises to do. I usually use SQLite for unit-tests and Postgres for integrated tests and production – my projects are filled with custom `@compiles` decorators to provide the same functionality across backends.

**Comment by zzzeek** [comment]
OK arent the differences for the typical regexp between a database like PG and SQLite mostly going to be the case insensitive part?   or are basic syntaxes within the regexp strings different between PG / SQLite  for example?

**Comment by zzzeek** [comment]
because we definitely can't be parsing and tokenizing regex strings, if the regex syntaxes are truly differnet and you wanted to make it so the user didn't have to know that, that's what that would entail, that's out of scope for SQLAlchemy (a third party extension could certainly do it however).

**Comment by jvanasco** [comment]
Flags- The multiline and newline modes are slightly different (see my chart)

Actual regex- 
Postgres - POSIX, Spencer library since 7.2.  TCL ARE by default, ERE and BRE supported by extending the string.
Mysql - POSIX ERE, but the backing library was Spencer until 8.0.4 and is now unicode
Sqlite - PCRE via plugin
Oracle - POSIX ERE

TLDR; they don't really support the same syntax. they can get close, but that involves changing the pattern. it's possible there is a session option.

**Comment by zzzeek** [comment]
I've used regular expressions for literally 25 years and i couldnt give you one difference between PCRE and POSIX, so I found this: https://gist.github.com/CMCDragonkai/6c933f4a7d713ef712145c5eb94a1816

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Implement regexp operator** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2092

**Comment by CaselIT** [comment]
Actually that gerrit is more the logic of with_variant

**Comment by CaselIT** [comment]
In the end it's easier to implement the with_variant construct for each boolean expression that making it a regexp only thing

**Comment by CaselIT** [comment]
Seems that maria db does not have a clear way of setting the flags that I could find, they are embedded in the pattern. https://mariadb.com/kb/en/pcre/#option-setting

Also it does not seem to be the case with mysql, that has https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-like and different flags

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Implement regexp operator** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2096

**Comment by CaselIT** [comment]
I've added a first implementation here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2096

to figure out:
- how to support mysql/mariadb flags, since they are not compatible with each other and we cannot discriminate them while compiling. mysql uses `regexp_like ` while mariadb is like postgres, end embeds them in the pattern.
- how to support flags for postgres, since they are embedded in the pattern. (mariadb has the same problem)

**Comment by CaselIT** [comment]
I've update the changset https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2096 to use `concat` for the flags in pg and issue a warning when flags are used in mysql.

Feedbacks welcome
